File: 15-3sum/3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(),nums.end());
        vector<vector<int>> result;
        int sizeNums = nums.size();
        for(int i=0; i<sizeNums; i++){
            if(i>0 && nums[i-1]==nums[i]) continue;

            int j = i+1;
            int k = sizeNums - 1;
            while(j<k){
                int sum = nums[i] + nums[j] + nums[k];
                if(sum==0){
                    result.push_back({nums[i], nums[j], nums[k]});
                    while(j<k && nums[k-1]==nums[k]) k--;
                    while(j<k && nums[j+1]==nums[j]) j++;
                    k--;
                    j++;
                }
                else if(sum > 0){
                    k--;
                }
                else{
                    j++;
                }            
            }
        }
        return result;
    }
};

File: 15-3sum/3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        result = []
        for i in range(len(nums)):
            if (i > 0 and nums[i] == nums[i-1]): continue

            j = i+1
            k = len(nums)-1
            while j<k:
                target = nums[i] + nums[j] + nums[k]
                if target == 0:
                    result.append([nums[i], nums[j], nums[k]])
                    while j<k and nums[j] == nums[j+1]: j+=1
                    while j<k and nums[k] == nums[k-1]: k-=1
                    j+=1
                    k-=1
                elif target > 0:
                    k-=1
                else:
                    j+=1
            
        return result